XenModel/Actions/CallHome/CallHomeAuthenticationAction.cs

```diff
@@ -63,31 +83,41 @@ protected override void Run()
             }
         }
 
+        public string UploadToken
+        {
+            get { return uploadToken; }
+        }
+
         public static void SetUploadTokenSecret(IXenConnection connection, Dictionary<string, string> config, string uploadToken)
         {
-            if (uploadToken == null)
+            SetTokenSecret(connection, config, CallHomeSettings.UPLOAD_TOKEN_SECRET, uploadToken);
+        }
+        
+        public static void SetTokenSecret(IXenConnection connection, Dictionary<string, string> config, string tokenKey, string tokenValue)
```

Comment:
I think it would be good to return when the tokenKey is null. (public static, may be reused later)